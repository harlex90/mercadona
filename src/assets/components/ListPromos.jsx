import caddie from "../../assets/caddie.jpg";
import guitare from "../../assets/guitare.jpg";
import fauteuil from "../../assets/fauteuil.jpg"
import tableau from "../../assets/tableau.jpg"



const products = [
    {
        "description": "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.",
        "id": 1,
        "image": guitare,
        "name": "Guitare bleu",
        "price": "1000€", 
        "new_price": "700€",
        "start_date": "4 mai 2023",
        "end_date": "1 juin 2023"
    }, {
        "description": "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.", 
        "id": 2, 
        "image": caddie, 
        "name": "Guitare rouge", 
        "price": "1100€",
        "new_price": "750€",
        "start_date": "30 avril 2023",
        "end_date": "12 juin 2023"
    }, {
        "description": "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.",
        "id": 1,
        "image": fauteuil,
        "name": "Guitare bleu",
        "price": "1000€", 
        "new_price": "700€",
        "start_date": "4 mai 2023",
        "end_date": "1 juin 2023"
    }, {
        "description": "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.", 
        "id": 2, 
        "image": tableau, 
        "name": "Guitare rouge", 
        "price": "1100€",
        "new_price": "750€",
        "start_date": "30 avril 2023",
        "end_date": "12 juin 2023"
    }, {
        "description": "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.",
        "id": 1,
        "image": guitare,
        "name": "Guitare bleu",
        "price": "1000€", 
        "new_price": "700€",
        "start_date": "4 mai 2023",
        "end_date": "1 juin 2023"
    }, {
        "description": "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.", 
        "id": 2, 
        "image": caddie, 
        "name": "Guitare rouge", 
        "price": "1100€",
        "new_price": "750€",
        "start_date": "30 avril 2023",
        "end_date": "12 juin 2023"
    }
];

const PromoCard = ({ product }) => {
    return(
        <div style={{display: "flex", flexDirection: "column"}}>
            <div style={{border: '1px solid black', borderRadius: '10px', width: "300px", height:"375px"}}> 
                <p style={{display: "flex", justifyContent: "center", fontSize:"20px"}}>{product.name}</p>
                <div style={{display: "flex", justifyContent: "center"}}>
                    <img
                        src={product.image}
                        alt={"product image"} 
                        style={{display: "flex", justifyContent: "center", width: 150, height: 150}}
                    />
                </div>
                <div style={{display: "flex", flexDirection: "column", marginRight:"10px", gap: "15px", fontSize:"20px"}}>
                    <p style={{display: "flex", justifyContent: "center", marginLeft: "10px", marginRight: "10px", marginBottom: "0px", fontSize:"15px"}}>{product.description}</p>
                    <div style={{display: "flex", justifyContent: "space-around", alignItems: "center"}}>
                        <div style={{display: "flex", flexDirection: "column"}}>
                            <p style={{fontSize:"12px", color: "red", margin: "0px"}}>En promotion du {product.start_date}</p>
                            <p style={{fontSize:"12px", color: "red", margin: "0px"}}>au {product.end_date}</p>
                        </div>
                        <div>
                            <p style={{textDecoration: "line-through", margin: "1px"}}>{product.price}</p>
                            <p style={{margin: "1px", color: "red", fontSize: "25px"}}>{product.new_price}</p>
                        </div>
                    </div>
                </div>
                
            </div>
        </div>
    )
}

const ListPromos = () => {
    return(
            <div style={{display: "flex", justifyContent: "space-around"}}>
                {products.map((product) => <PromoCard product={product} />)}
            </div>
    )
}

export default ListPromos;