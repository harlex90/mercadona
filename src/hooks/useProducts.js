
import React from 'react';
import axios from "axios";

const useProducts = () => {
    const [isLoading, setIsLoading] = React.useState(true);
    const [products, setProducts] = React.useState([]);

    React.useEffect(() => {
        axios.get('/products').then((res) => {
            setProducts(res.data);
            setIsLoading(false);
        })
    }, []);

    return {
        products,
        isLoading,
    };
}

export default useProducts;