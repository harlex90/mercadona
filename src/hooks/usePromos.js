
import React from 'react';
import axios from "axios";

const usePromos = () => {
    const [isLoading, setIsLoading] = React.useState(true);
    const [promos, setPromos] = React.useState([]);

    React.useEffect(() => {
        axios.get('/promotions').then((res) => {
            setPromos(res.data);
            setIsLoading(false);
        })
    }, []);

    return {
        promos,
        isLoading,
    };
}

export default usePromos;