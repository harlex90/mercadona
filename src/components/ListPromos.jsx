import moment from 'moment';
import usePromos from '../hooks/usePromos';
import useProducts from '../hooks/useProducts';
import { groupBy } from 'lodash';
import findMaxPromo from '../assets/utils/findMaxPromo';
import ItemCard from './ItemCard';


const ListPromos = () => {
    const { products } = useProducts();
    const { promos: allPromos } = usePromos();


    const now = Date.now() / 1000; // TODO remove /1000 when backend fixed
    const validPromos = allPromos.filter((prom) => {
        return (prom.start_date < now && prom.end_date > now)
    });

    const promosByProducts = groupBy(validPromos, (promo) => promo.product_id);
    const productsPromos = Object.keys(promosByProducts).map((key) => {
       const promosByProduct = promosByProducts[key];
       const betterDiscount = findMaxPromo(promosByProduct);
       return betterDiscount;
    });

    const items = [];
    productsPromos.forEach((promo) => {
        const product = products.find((prod) => {
            return (prod.id === promo.product_id)
        })
        if(!product) return null;

        const startDate = moment.unix(promo.start_date)
        const endDate = moment.unix(promo.end_date)
        const today = moment()
        const isBetween = today.isSameOrAfter(startDate, 'days') && today.isSameOrBefore(endDate, 'days')
        if(!isBetween) return null

        items.push({
            ...promo,
            ...product,
            new_price: product.price * (1 - promo.discount),
            id: product.id,
        })
    });

    return(
        <div style={{display: "flex", justifyContent: "space-around", flexWrap: "wrap"}}>
            {items.map((product) => <li key={product.id} style={{listStyle: "none"}}><ItemCard product={product} /></li>)}
        </div>
    )
}

export default ListPromos;
