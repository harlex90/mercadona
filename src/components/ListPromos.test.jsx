import { render } from '@testing-library/react';
import ListPromos from './ListPromos';
import usePromos from '../hooks/usePromos';
import useProducts from '../hooks/useProducts';

jest.mock('react-auth-kit', () => ({
    useIsAuthenticated: () => () => false, // later, we will change implementation of this mock BY TEST
    useSignOut: () => () => console.log('signOut()')
}))

const mockNavigate = jest.fn();
jest.mock('react-router-dom', () => ({
    useNavigate: () => mockNavigate
}))

jest.mock('./ItemCard', () => ({ product }) => {
    return <div data-testid="ItemCard" data-product={product.id} data-discount={product.discount } />
})

jest.mock("../hooks/usePromos", () => jest.fn())
jest.mock("../hooks/useProducts", () => jest.fn())

describe('ListPromos', () => {
    beforeEach(() => {
        useProducts.mockImplementation(() => {
            return ({
                products: []
            })
        })
        usePromos.mockImplementation(() => {
            return ({
                promos: []
            })
        })
    });

    test('match snapshot', () => {
        const { asFragment } = render(<ListPromos />);
        expect(asFragment()).toMatchSnapshot()
    });

    test('match snapshot when data', () => {
        useProducts.mockImplementation(() => {
            return ({
                products: [{
                    id: 3,
                    name: 'chaise',
                    category_id: 9,
                }]
            })
        })
        usePromos.mockImplementation(() => {
            return ({
                promos: [{
                    id: 5,
                    product_id: 3,
                    start_date: 0,
                    end_date: 10111111111,
                    discount: 0.5
                }]
            })
        })
        
        const { asFragment } = render(<ListPromos />);
        expect(asFragment()).toMatchSnapshot()
    });
    test('should display startDate and endDate which are before now and efter now', () => {
        // const startDate = prom.start_date
        // const endDate = prom.end_date
        // expect(startDate).toBeLessThan(mockDate.now)
        // expect(endDate).toBeGreaterThan(mockDate.now)
        const { asFragment } = render(<ListPromos />);
        
    });
});

                