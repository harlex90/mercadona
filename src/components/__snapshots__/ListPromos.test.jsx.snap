// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListPromos match snapshot 1`] = `
<DocumentFragment>
  <div
    style="display: flex; justify-content: space-around; flex-wrap: wrap;"
  />
</DocumentFragment>
`;

exports[`ListPromos match snapshot when data 1`] = `
<DocumentFragment>
  <div
    style="display: flex; justify-content: space-around; flex-wrap: wrap;"
  >
    <li
      style="list-style: none;"
    >
      <div
        data-discount="0.5"
        data-product="3"
        data-testid="ItemCard"
      />
    </li>
  </div>
</DocumentFragment>
`;
